### PR TITLE
Implement generic forwarder

### PR DESCRIPTION
**Pull request by Blinkuu**
**What this PR does**:

This PR introduces a new subcomponent to `distributor` named `forwarder`. This generic forwarder serves as a facade for forwarding batches of spans in "fire-and-forget" way. The feature is designed in such a way that it's configurable on per-tenant basis. 

Example configuration:
```yaml
# config.yaml
distributor:
  forwarders:
    - name: "otel-forwarder"
      backend: "otlpgrpc"
      otlpgrpc:
        endpoints: ['otelcol:4317']
        tls:
          insecure: true

# overrides.yaml
overrides:
  "example-tenant-1":
    forwarders: ['otel-forwarder']
  "example-tenant-2":
    forwarders: ['otel-forwarder']
```

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by mapno — COMMENTED** [review]
Gave it a quick review. I will come back to it later, it's a big PR. Overall I like the direction of the PR.

File: modules/distributor/queue/queue.go
Comment (Blinkuu) [reply modules/distributor/queue/queue.go:106]: The core idea was to remove `context.Background` that was previously used in queue manager, like this:

```go
m.forwardRequest(context.Background(), req)
```

and add at least some timeout. I've put it here because it's now easy to derive this context from the context that we receive in Push() as an argument if need be. I don't have strong opinions on this, though.

Comment (Blinkuu) [reply modules/distributor/queue/queue.go]: I know understand what you meant when you wrote "context is canceled". You're right, that's not the intended behaviour. I fixed it and moved the context back where it was before (to the queue itself). However, I kept the 10s timeout :)

**Review by mapno — COMMENTED** [review]
Overall, looks very good to me. I only find the manager logic to be complicated and difficult to follow. Left a comment.

Will wait to see if anyone else wants to chip in.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:157]: Yeah, I was wondering the same. Tbh, I didn't include WithBlock() as the default option, since I couldn't find it anywhere in the codebase and thought that this is the default you prefer. I personally prefer WithBlock(), as I'm a strong believer that APIs in go should be blocked by default :P Will add this :)

Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:158]: Added in [c946376](https://github.com/grafana/tempo/pull/1775/commits/c946376c740cc37bec99a9c5f484ca5aad12d7c0)

File: modules/distributor/distributor.go
Comment (joe-elliott) [inline modules/distributor/distributor.go:367]: we should metric this in a way we can expose to grafana cloud tracing customers. `tempo_warnings_total` is already piped back to the user and was meant to be a place for generic warnings.

File: modules/distributor/distributor.go
Comment (joe-elliott) [inline modules/distributor/distributor.go:365]: nit: seems unnecessary to check len of `d.cfg.Forwarders`

File: modules/distributor/forwarder.go
Comment (joe-elliott) [inline modules/distributor/forwarder.go:88]: this would be far easier to review if we had split the refactor from the addition of the new forwarder. is that still possible?

**Review by joe-elliott — COMMENTED** [review]
This is a large PR that will require several hours to give a proper review to. Is there anyway to break the refactor up from the functionality addition?

File: modules/distributor/distributor.go
Comment (Blinkuu) [reply modules/distributor/distributor.go:365]: len(d.cfg.Forwarder) can be 0 in a situation where there is a forwarder defined for a particular tenant in the `overrides.yaml` config, but forwarders are defined in the distributor's `config.yaml`.

Also, this check serves as an additional safety net, it ensures that the feature is completely toggled off when there are no forwarders defined in distributor's `config.yaml`. If there were a bug or anything, it guarantees a super simple way to shut down this feature.

File: modules/distributor/distributor.go
Comment (Blinkuu) [reply modules/distributor/distributor.go:367]: Frankly, I don't understand. Do you mean that I cannot use warn-level logs because there is some random side-effect Could you provide some more context?

**Comment by Blinkuu** [comment]
This PR is now rebased on top of #1796. That PR *must* be merged before this one.

File: modules/distributor/forwarder.go
Comment (Blinkuu) [reply modules/distributor/forwarder.go:88]: Done.

1. Refactor: #1796
2. I've rebased this PR on top of #1796. It will become easier to review after we merge #1796 to origin/main.

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:89]: let's back this up a bit. 10 seconds? 15 seconds?

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:140]: i feel like this is unnecessarily complex. why don't we just iterate through `m.tenantToQueueList`. if the tenant is not in `tenantToForwaderNames` then call remove?

we can then call update on anything remaining.

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:193]: if this method is only called from `ForTenant()` then why don't we remove it and place the contents in `ForTenant()`?

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:215]: instead of individually deleting create a new, empty map after the loop?

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:223]: same as above

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:33]: is this used? if not remove

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:236]: is this used? if not remove

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [inline modules/distributor/forwarder/manager.go:268]: similar to above there is likely a more straightforward way to write this.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (joe-elliott) [inline modules/distributor/forwarder/otlpgrpc/forwarder.go:79]: we really need to avoid this marshal/unmarshal dance. we already do this once in the distributor and i've been trying to find a way to remove that for years.

since our trace is wire compatible with otlp we can just marshal it once when we send it and the other side should able to unmarshal.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (joe-elliott) [inline modules/distributor/forwarder/otlpgrpc/forwarder.go:95]: this is a lot of goroutines at scale. this needs to send synchronously and rely on the queue to handle slowdowns

**Review by joe-elliott — COMMENTED** [review]
I think we can reduce the complexity of this PR significantly. Let me know if some of these comments are unclear.

File: modules/distributor/distributor.go
Comment (zalegrala) [reply modules/distributor/distributor.go:367]: If we don't add a metric to this PR, it might be handy to have an issue to make sure we come back to it.  A distributor metric might be handy for us internally too.

File: modules/distributor/forwarder/config.go
Comment (Blinkuu) [reply modules/distributor/forwarder/config.go]: Removed

File: modules/distributor/forwarder/forwarder.go
Comment (Blinkuu) [reply modules/distributor/forwarder/forwarder.go:20]: I don't think the `List` type should implement the `Forwarder` interface in its entirety. This type is just for convenience.

The underlying forwarders are basically queues (`[]queue.Queue` wrapped in a `queueAdapter` to implement a `Forwarder` interface). The shutdown method is called on them in a couple of scenarios:

1.  `forwarder.Manager` is stopped. This triggers a call to `Manager.shutdown()` method which calls shutdown on all underlying queues via a `queueList.shutdown()`. Additionally, `Manager.shutdown()` will also call shutdown on the real forwarders.
2. queueList gets deleted when a configuration for a particular tenant is removed. This happens in `Manager.removeQueueListForTenantUnderLock()`.
3. queueList gets updated when new configuration removes a reference to a forwarder for a particular tenant. This happens in `queueList.update()` while iterating over `diff.removed` slice.

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Removed, was used before.

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Changed to 10 seconds.

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Removed

File: modules/distributor/forwarder/otlpgrpc/config.go
Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/config.go]: Removed

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go]: Made is synchronous

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:79]: The problem is that the `Export()` method of `ptraceotlp.Client` accepts a `ptraceotlp.Request` as an argument, and I don't think there is a simple way to just convert `tempopb.Trace` type directly into the request type. I'd rather not do some reinterpret cast magic.

I don't know enough about Go, but in C++ such stuff would most likely be optimized away :D

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go:75]: We cannot return the original slice. First of all, there is no "original" slice. `list.copy()` constructs a slice from a map. Secondly, the underlying map can (and will) change on overrides configuration updates.

If we want to get rid of those allocations, perhaps the `ForTenant()` could be changed to `ForwardBatchesForTenant()` and we would just call forward batches on each queue under a lock.

Let me know your thoughts.

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go:137]: I don't think it's that simple. This solution does not cover new tenant IDs showing up in tenantToForwarderNames. To be honest, I like the "diff" solution. Perhaps it's a little more code than if I were to just iterate over those two maps, but IMO it's very descriptive.

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go:190]: I see no value in removing this function

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Done

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Done

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go:263]: Yeah, similar comment here.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (mapno) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:79]: You can build the request with `ptraceotlp.NewRequestFromTraces(td)` if the conversion is moved up from the [shim](https://github.com/grafana/tempo/blob/main/modules/distributor/receiver/shim.go#L270-L283) to the distributor's `PushBatches`. The downside is that we would leak OTel to the distributor, but IMO it's worth for avoiding the conversion.

Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:79]: I don't have a strong opinion on this. @joe-elliott which option would you prefer?

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (mapno) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go]: Optionally, we could use a [boundedwaitgroup](https://github.com/grafana/tempo/tree/main/pkg/boundedwaitgroup) per forwarder

Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go]: I think keeping it synchronous is good enough. On top of that, I think this feature would be primarily used in "single forwarder-single endpoint" manner. Defining multiple endpoints is kind of an edge case.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (mapno) [inline modules/distributor/forwarder/otlpgrpc/forwarder.go:118]: Now that sending traces is sync we could remove this fn for simplicity

File: modules/distributor/forwarder/forwarder.go
Comment (joe-elliott) [reply modules/distributor/forwarder/forwarder.go:20]: i'm fine with this. just noticed it partially implemented the interface and wondered if you saw value in doing the whole thing.

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [reply modules/distributor/forwarder/manager.go:190]: less code?

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [reply modules/distributor/forwarder/manager.go:137]: I propose this instead:

- queueList is immutable. it cannot be updated (cleans up all that code)
- on update of forwarder manager
  - loop through all existing queueLists. remove all that are to be deleted and all that need to be updated
  - call shutdown on all of these queueLists (removed and updated). this will cause them to properly drain async.
  - readd new queueLists for all updated ones that will be used by future pushes

this will clean up all the diffing code and all of the queueList update code. It will also simplify this code significantly. Since queue lists become immutable you can likely also remove all locking code there.

File: modules/distributor/forwarder/manager.go
Comment (joe-elliott) [reply modules/distributor/forwarder/manager.go:75]: Making the queueList immutable as described below opens up the possibility to return a slice owned by the queueList for reduced allocations. I'll admit I'm not honestly sure what the impact will be here. It will certainly be far less then the proto marshalling discussion below.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (joe-elliott) [inline modules/distributor/forwarder/otlpgrpc/forwarder.go:51]: i would really like to get rid of the locking code in this object. i think we can do that using the concurrency controls in the forwarder manager one package up. basically, making sure that lifecycle events Dial, ForwardBatches, and Shutdown are never called concurrently. it may already be true.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (joe-elliott) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:79]: > I'd rather not do some reinterpret cast magic.
100% agreed

I like @mapno's solution here.

File: modules/distributor/forwarder/manager.go
Comment (mapno) [reply modules/distributor/forwarder/manager.go:137]: I like this proposal 👍 It's very close to the logic from the original forwarder, so it makes sense to me.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:51]: I wouldn't change this. The idea is that many tenants can "attach" to a single Forwarder instance. The flow is kind of like this:

Distributor -> Queue (per tenant) -> Forwarder

Many tenants share the same forwarder. Also, the Dial is just called once, during Tempo startup, and `ForwardBatches` only acquires a read-lock and does so on Queue's thread (in workers, essentially). I don't think we have much to gain with this micro-optimization.

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (Blinkuu) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go]: Done in [2de3a5c](https://github.com/grafana/tempo/pull/1775/commits/2de3a5c157ebb23eec7d1f21af567751ace2b6ef).

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Done in [1871143](https://github.com/grafana/tempo/pull/1775/commits/187114373c152431a9ecdda8f4cf9136d3e81cca).

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Done in [57fde51](https://github.com/grafana/tempo/pull/1775/commits/57fde5127c307c25414c4c5eb2bfe6afeeb1f818).

`queueList` is now immutable

**Review by mapno — COMMENTED** [review]
I think it's almost there. Great work 👏

File: modules/distributor/forwarder/manager.go
Comment (Blinkuu) [reply modules/distributor/forwarder/manager.go]: Done in [e94a896](https://github.com/grafana/tempo/pull/1775/commits/e94a896103fdae5d7a4d24b9d0c5511c3f237eef).

File: modules/distributor/distributor.go
Comment (joe-elliott) [inline modules/distributor/distributor.go:234]: ultimately this calls `grpc.DialContext()` which (i believe) attempts an asynchronous tcp connection. if so, then the code is fine as is.

However, if `grpc.DialContext()` blocks and returns an error on an unsuccessful connection then we need to make changes. i.e. If a forwarder endpoint is down then it is important that a distributor will start successfully.

Can you confirm the behavior either way?

File: modules/distributor/forwarder/otlpgrpc/forwarder.go
Comment (joe-elliott) [reply modules/distributor/forwarder/otlpgrpc/forwarder.go:49]: For me, it's less about optimization and more about only using locks where necessary to reduce code complexity. However, I consider this minor am fine with leaving it in.

**Review by joe-elliott — COMMENTED** [review]
One final question and then I'm good.

File: modules/distributor/distributor.go
Comment (Blinkuu) [reply modules/distributor/distributor.go:234]: Done in [28e63a1](https://github.com/grafana/tempo/pull/1775/commits/28e63a15e5861ff2f40c09d7f48aacca158abec7).

**Review by mapno — APPROVED** [review]
LGTM

**Pull request merged**
